File: Leetcode Daily Challenge/026-August-2025/004-fruit-into-baskets.java
// https://leetcode.com/problems/fruit-into-baskets/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(N)
    public int totalFruit(int[] fruits) {
        int n = fruits.length;
        Map<Integer, Integer> mpp = new HashMap<>();
        int i = 0, j = 0, maxFruits = 0;
        while (j < n) {
            mpp.put(fruits[j], mpp.getOrDefault(fruits[j], 0) + 1);
            while (mpp.size() > 2) {
                mpp.put(fruits[i], mpp.get(fruits[i]) - 1);
                if (mpp.get(fruits[i]) == 0) {
                    mpp.remove(fruits[i]);
                }
                i ++;
            }
            maxFruits = Math.max(maxFruits, j - i + 1);
            j ++;
        }
        return maxFruits;
    }
}

File: Leetcode Daily Challenge/026-August-2025/004-fruit-into-baskets.cpp
// https://leetcode.com/problems/fruit-into-baskets/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    // TC : O(N)
    // SC : O(1) only store at max 2-elements
    int approach_1(vector<int> &fruits) {
        int n = fruits.size();
        unordered_map<int, int> mpp;
        int i = 0, j = 0, maxFruit = 0;
        while (j < n) {
            mpp[fruits[j]] ++;
            while (mpp.size() > 2) {
                mpp[fruits[i]] --;
                if (!mpp[fruits[i]]) mpp.erase(fruits[i]);
                i ++;
            }
            maxFruit = max(maxFruit, j - i + 1);
            j ++;
        }
        return maxFruit;
    }
public:
    int totalFruit(vector<int>& fruits) {
        return approach_1(fruits);
    }
};